### PR TITLE
Rework the reference cache

### PR DESCRIPTION
**Pull request by benashz**
Previously, the cache was keyed by referent object. This made things simpler for referent management, but it came at the cost of extra memory usage and some added code complexity.

In order to mitigate the potential for encountering an out of memory situation, the cache will map a referring object to its set of references.

Other fixes:
- The previous implementation only supported adding or pruning referrers, which would lead to orphaned references persisting in the cache. So, Add() is replaced by Set(), which sets the set of references for each referring object.
- Make ReferenceCache a private member on all secret controllers.

### CODE REVIEW

File: controllers/hcpvaultsecretsapp_controller.go [hunk 1]
```diff
@@ -128,16 +128,10 @@ func (r *HCPVaultSecretsAppReconciler) Reconcile(ctx context.Context, req ctrl.R
...
-			r.ReferenceCache.Add(SecretTransformation, ref,
-				req.NamespacedName)
-		}
-	} else {
-		r.ReferenceCache.Prune(SecretTransformation, req.NamespacedName)
-	}
+	r.referenceCache.Set(SecretTransformation, req.NamespacedName,
+		helpers.GetTransformationRefObjKeys(
+			o.Spec.Destination.Transformation, o.Namespace)...)
+	logger.Info("ReferenceCache", "cache", r.referenceCache)
```
Comment (tvoran) [inline controllers/hcpvaultsecretsapp_controller.go:134]: Should this be debug or trace?

**Review by tvoran — APPROVED** [review]
I think I get it, though I'm not sure I follow how this changes the memory usage. Can you elaborate on that?

It also might help to have more concrete examples of what the objects are in the doc strings; I'm finding it hard to keep referent/referrer/reference straight.

File: controllers/eventhandlers.go [hunk 2]
```diff
@@ -92,33 +92,28 @@ func (e *enqueueRefRequestsHandler) enqueue(ctx context.Context,
...
-			return
-		}
 
+	for _, ref := range e.refCache.Get(e.kind, client.ObjectKeyFromObject(o)) {
 		if e.validator != nil {
 			if err := e.validator(ctx, o); err != nil {
 				logger.Error(err, "Validation failed, skipping enqueue")
 				return
 			}
 		}
```
[the remark this answers is not in the document]
Comment (benashz) [reply controllers/eventhandlers.go:102]: True, but I think that we only need to validate if have any referring client.ObjectKeys.  Validating the referenced object could be a fairly expensive. We can probably make it conditional on `len(referrers) > 0`

File: controllers/hcpvaultsecretsapp_controller.go (hunk 1 above)
Comment (benashz) [reply controllers/hcpvaultsecretsapp_controller.go:134]: Removed in:

File: controllers/eventhandlers.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (benashz) [reply controllers/eventhandlers.go:102]: Updated in:

**Comment by benashz** [comment]
> I think I get it, though I'm not sure I follow how this changes the memory usage. Can you elaborate on that?

It avoids duplicating the referring object e.g VaultStaticSecrets, since the mapping is `secret object -> [SecretTransformation,...]`. Typically the number of SecretTransformation instances should be lower than the number of secret objects configured in a cluster.

**Comment by benashz** [comment]
> It also might help to have more concrete examples of what the objects are in the doc strings; I'm finding it hard to keep referent/referrer/reference straight.

Should be addressed in 3449d087d69def2dabf115b03c6f64669f8d0953

**Pull request merged**